base/src/Data/Macaw/Analysis/RegisterUse.hs

```diff
@@ -567,53 +1658,151 @@ backPropagateOne s rest newLocs ((srcAddr,srcDepMap):predRest) = do
+biAssignIdUsed aid inv = Set.member (Some aid) (biUsedAssignSet inv)
+
+-- | Return true if index corresponds to a write of the current stack
+-- frame.
+biWriteUsed :: StmtIndex -> BlockInvariants arch ids -> Bool
+biWriteUsed idx inv = Set.member idx (biUsedWriteSet inv)
+
 -- | This transitively back propagates blocks across
-backPropagate :: forall r ids
-              .  (MapF.OrdF r, MemWidth (RegAddrWidth r))
-              => PredProvideMap r ids
+backPropagate :: forall arch ids
+              .  (OrdF (ArchReg arch), MemWidth (ArchAddrWidth arch))
+              => PredProvideMap (ArchReg arch) ids
               -- ^ Pred provide map computed during summarization.
-              -> FixState r ids
-              -> NewDemandMap r
-              -- ^ New demands on block addresses.
-              -> FixState r ids
-backPropagate predMap s new =
+              -> Map (ArchSegmentOff arch)  (DependencySet (ArchReg arch) ids)
+              -> Map (ArchSegmentOff arch) (Set (Some (BoundLoc (ArchReg arch))))
+              -- ^ Maps block addresses to the set of locations that
+              -- we still need to back propagate demands for.
+              -> Map (ArchSegmentOff arch)  (DependencySet (ArchReg arch) ids)
+backPropagate predMap depMap new =
   case Map.maxViewWithKey new of
-    Nothing -> s
+    Nothing -> depMap
     Just ((currAddr, newRegs), rest) ->
       let predAddrs = Map.findWithDefault [] currAddr predMap
-          (s', rest') = backPropagateOne s rest newRegs predAddrs
+          (s', rest') = backPropagateOne depMap rest newRegs predAddrs
        in backPropagate predMap s' rest'
 
 ------------------------------------------------------------------------
 -- registerUse
 
+-- | Create map from locations to the non-representative locations
+-- that are equal to them.
+mkDepLocMap :: forall arch
+            .  OrdF (ArchReg arch)
+            => BlockStartConstraints arch
+            -> MapF (BoundLoc (ArchReg arch)) (LocList (ArchReg arch))
+mkDepLocMap cns =
+  let addNonRep :: MapF (BoundLoc (ArchReg arch)) (LocList (ArchReg arch))
+                -> BoundLoc (ArchReg arch) tp
+                -> ValueRegUseDomain arch tp
+                -> MapF (BoundLoc (ArchReg arch)) (LocList (ArchReg arch))
+      addNonRep m l (RegEqualLoc r) = MapF.insertWith (<>) r (LL [l]) m
+      addNonRep m _ _ = m
+   in foldLocMap addNonRep MapF.empty (bscLocMap cns)
+
+mkBlockInvariants :: forall arch ids
+                  .  (HasRepr (ArchReg arch) TypeRepr
+                     , OrdF (ArchReg arch)
+                     , ShowF (ArchReg arch)
+                     , MemWidth (ArchAddrWidth arch)
+                     )
+                  => FunPredMap (ArchAddrWidth arch)
+                  -> (ArchSegmentOff arch
+                       -> ArchSegmentOff arch
+                       -> PostValueMap arch ids)
+                     -- ^ Maps block address to map of location representative to value.
```

Comment:
Is the comment implying that I should think of this as a function where I apply a block address to get a (map of location representatives to values)?